velox/functions/prestosql/GeometryUtils.h

```diff
@@ -0,0 +1,685 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <geos/geom/GeometryFactory.h>
+#include <geos/geom/LineString.h>
+#include <geos/geom/Point.h>
+
+#include <velox/common/base/Exceptions.h>
+#include <velox/type/StringView.h>
+
+namespace facebook::velox::functions {
+
+class GeometryUtils {
+ public:
+  template <typename StringWriter>
+  static void serialize(
+      const std::unique_ptr<geos::geom::Geometry>& geometry,
+      StringWriter& stringWriter) {
+    SliceOutput sliceOutput(stringWriter);
+    serialize(geometry.get(), sliceOutput);
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> deserialize(
+      const StringView& geometryString) {
+    SliceInput sliceInput(geometryString.data(), geometryString.size());
+    return deserialize(sliceInput);
+  }
+
+  static std::unique_ptr<geos::geom::Geometry> createPoint(double x, double y) {
+    geos::geom::Coordinate coord(x, y);
+    return std::unique_ptr<geos::geom::Point>(
+        GEOMETRY_FACTORY->createPoint(coord));
+  }
+
+ private:
+  enum class GeometrySerializationType : uint8_t {
+    POINT = 0,
+    MULTI_POINT = 1,
+    LINE_STRING = 2,
+    MULTI_LINE_STRING = 3,
+    POLYGON = 4,
+    MULTI_POLYGON = 5,
+    GEOMETRY_COLLECTION = 6,
+    ENVELOPE = 7
+  };
+
+  enum class EsriShapeType : int {
+    POINT = 1,
+    POLYLINE = 3,
+    POLYGON = 5,
+    MULTI_POINT = 8
+  };
+
+  class SliceInput {
```

Comment:
Could you use https://github.com/facebookincubator/velox/blob/main/velox/common/base/IOUtils.h instead?  This seems like it's re-implementing that.

Reply:
That makes sense. Yeah, we probably don’t need the boundary check. As for serialization, I initially implemented it the same way, but then I refactored it to avoid the extra copy. If you'd prefer the original approach, I can revert the code back to that.

Reply:
@kgpai I agree that for deserialization, boundary checks aren't necessary. However, as I mentioned earlier, for serialization we can't predict the final size in advance, so pre-allocating space isn't feasible. While it's possible to use `InputByteStream`, we can't directly use `OutputByteStream` in this case. Do you have any suggestions or alternatives we could consider?

Reply:
> @kgpai I agree that for deserialization, boundary checks aren't necessary. However, as I mentioned earlier, for serialization we can't predict the final size in advance, so pre-allocating space isn't feasible. While it's possible to use `InputByteStream`, we can't directly use `OutputByteStream` in this case. Do you have any suggestions or alternatives we could consider?

@wraymo This approach makes sense to me. Let's switch to using InputByteStream for deserialization and at the end we can do a sanity check for bounds by comparing offset() to the size of the input (this should be cheap as its only done once per input). For serialization, SliceOutput works well, but I have a minor suggestion: could we rename it to something like `VarbinaryWriter`?